File: java/src/main/java/problems101_200/ValidPalindrome.java
package problems101_200;

enum ValidPalindrome {
    REVERSE_STRING {
        @Override
        boolean isPalindrome(String s) {
            StringBuilder builder = new StringBuilder();
            for (char c : s.toCharArray()) {
                if (Character.isLetterOrDigit(c)) {
                    builder.append(c);
                }
            }

            return builder.toString().toLowerCase()
                .contentEquals(builder.reverse().toString().toLowerCase());
        }
    },
    COMPARE_HALF_STRING {
        @Override
        boolean isPalindrome(String s) {
            StringBuilder builder = new StringBuilder();
            for (char c : s.toCharArray()) {
                if (Character.isLetterOrDigit(c)) {
                    builder.append(c);
                }
            }

            s = builder.toString();
            int start = 0;
            int end = s.length() - 1;
            while (start < end) {
                if (Character.toLowerCase(s.charAt(start++))
                    != Character.toLowerCase(s.charAt(end--))) {
                    return false;
                }
            }
            return true;
        }
    };

    abstract boolean isPalindrome(String s);
}

File: kotlin/src/main/kotlin/problems101_200/ValidPalindrome.kt
package problems101_200

import java.util.Locale

enum class ValidPalindrome {
    REVERSE_STRING {
        override fun isPalindrome(s: String): Boolean {
            val builder = StringBuilder()
            for (c in s.toCharArray()) {
                if (Character.isLetterOrDigit(c)) {
                    builder.append(c)
                }
            }

            return builder
                .toString()
                .lowercase(Locale.getDefault())
                .contentEquals(builder.reverse().toString().lowercase(Locale.getDefault()))
        }
    },
    COMPARE_HALF_STRING {
        override fun isPalindrome(s: String): Boolean {
            var s = s

            val builder = StringBuilder()
            for (c in s.toCharArray()) {
                if (Character.isLetterOrDigit(c)) {
                    builder.append(c)
                }
            }

            s = builder.toString()
            var start = 0
            var end = s.length - 1
            while (start < end) {
                if (s[start++].lowercaseChar() != s[end--].lowercaseChar()) {
                    return false
                }
            }
            return true
        }
    },
    ;

    abstract fun isPalindrome(s: String): Boolean
}
